nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol

```diff
@@ -258,6 +284,109 @@ contract StakingEscrow is Upgradeable, IERC900History {
+    */
+    function withdraw(uint256 _value) external onlyStaker {
+        require(_value > 0, "Value must be specified");
+        StakerInfo storage info = stakerInfo[msg.sender];
+        require(info.flags.bitSet(MERGED_INDEX), "Merge must be confirmed");
+        require(
+            _value + tStaking.stakedNu(info.operator) <= info.value,
+            "Not enough tokens unstaked in T staking contract"
+        );
+        require(
+            _value + getVestedTokens(msg.sender) <= info.value,
+            "Not enough tokens released during vesting"
+        );
+        info.value -= _value;
+
+        token.safeTransfer(msg.sender, _value);
+        emit Withdrawn(msg.sender, _value);
+    }
+
+    /**
+    * @notice Returns amount of not released yet tokens for staker
+    */
+    function getVestedTokens(address _staker) public view returns (uint256) {
+        StakerInfo storage info = stakerInfo[_staker];
+        if (info.vestingReleaseTimestamp <= block.timestamp) {
+            return 0;
+        }
+        uint256 vestedTokens = (info.vestingReleaseTimestamp - block.timestamp) * info.vestingReleaseRate;
+        return info.value < vestedTokens ? info.value : vestedTokens;
+    }
+
+    /**
+    * @notice Setup vesting parameters
+    * @param _stakers Array of stakers
+    * @param _releaseTimestamp Array of timestamps when stake will be released
+    * @param _releaseRate Array of release rates
+    */
+    function setupVesting(
+        address[] calldata _stakers,
+        uint256[] calldata _releaseTimestamp,
+        uint256[] calldata _releaseRate
+    ) external onlyOwner {
+        require(_stakers.length == _releaseTimestamp.length &&
+            _releaseTimestamp.length == _releaseRate.length,
+            "Input arrays must have same number of elements"
+        );
+        for (uint256 i = 0; i < _stakers.length; i++) {
+            address staker = _stakers[i];
+            StakerInfo storage info = stakerInfo[staker];
+            require(info.vestingReleaseTimestamp == 0, "Vesting parameters can be set only once");
+            info.vestingReleaseTimestamp = _releaseTimestamp[i];
+            info.vestingReleaseRate = _releaseRate[i];
+            require(getVestedTokens(staker) > 0, "Vesting parameters must be set properly");
+            emit VestingSet(staker, info.vestingReleaseTimestamp, info.vestingReleaseRate);
+        }
+    }
+
+    /**
+    * @notice Request migration to threshold network
+    * @param _staker Staker address
+    * @param _operator Operator address
+    * @return Amount of tokens
+    */
+    function requestMerge(address _staker, address _operator)
+        external onlyTStakingContract returns (uint256)
+    {
+        StakerInfo storage info = stakerInfo[_staker];
+        require(
+            info.operator == address(0) ||
+            info.operator == _operator ||
+            (tStaking.stakedNu(info.operator) == 0 &&
+            !info.flags.bitSet(MERGED_INDEX)),
```

Comment:
The point of this is to make it more equitable between those who have vested tokens and those who have staked but non-vested tokens, by restricting the latter's ability to re-request a merge with a new operator address? That makes sense, particularly as doing so may have implications for their eligibility to authorize to tbtcv2 later. In general the more uniform the action space is for stakers (with different situations) the better, mostly to avoid irreversible moves where the benefits are massively outweighed by the costs, but the staker isn't aware of those downsides.

Reply:
Not really. The thing is after `confirmMerge` staker can't restake and change oeprator without withdrawing all NU. So for non-vested stakers it's easy (because they can withdraw everything in that case), but for vested-stakers it won't be possible before everything is unlocked. So the solution (for vested stakers) is - do not call `confirmMerge` before you're sure in operator address. And my question related to: it's not necessery to restrict like that (only before `confirmMerge`) but I have itchy feeling not to do that